core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1117,7 +1126,8 @@ private void ensureWriteAllowed(Engine.Operation op) throws IllegalIndexShardSta
 
     private void verifyPrimary() {
         if (shardRouting.primary() == false) {
-            throw new IllegalStateException("shard is not a primary " + shardRouting);
+            // TODO throw a more appropriate exception
+            throw new ShardNotFoundException(shardRouting.shardId(), "shard is not a primary anymore");
```

Comment:
why did you changed it? I think it's good to keep it consistent with verifyReplicationTarget and not use an exception that is covered by isShardNotAvailableException. These days, with aId checks and all, end up here by mistake is something we shouldn't swallow.